### ISSUE TITLE
[Question] Adding markers for each env in IsaacLab Manager based RL

### ISSUE THREAD

**Issue by shiven2001** [isaac-sim/IsaacLab#2779]
Hi, I wanted to ask how I can add markers for each env in my manager based RL env. I used the interactive scene and as mentioned in a previous post and added markers in post innit. It works but as u can see, the markers relative to world. I want to spawn the markers for each env relative to the env coordinates but when I set prim path as {ENV_REGEX_NS}/Visuals/mymarkers, it says prim doesn't exist....

Here is the code:

      @configclass
      class SceneCfg(InteractiveSceneCfg):
          """Configuration for a Wheeled robot scene."""
      
          # environment
          environment = AssetBaseCfg(
              prim_path="{ENV_REGEX_NS}/environment",
              spawn=sim_utils.UsdFileCfg(usd_path=usd_path),
              init_state=AssetBaseCfg.InitialStateCfg(pos=(0.0, 0.0, 0.0)),
          )
      
          # wall
          wall = AssetBaseCfg(
              prim_path="{ENV_REGEX_NS}/environment/wall",
              spawn=None,
              init_state=AssetBaseCfg.InitialStateCfg(),
          )
      
          # float_cam
          float_cam = RigidObjectCfg(
              prim_path="{ENV_REGEX_NS}/environment/float_cam",
              spawn=None,
              init_state=RigidObjectCfg.InitialStateCfg(),
          )
      
          # float_cam_cam
          float_cam_cam = CameraCfg(
              prim_path="{ENV_REGEX_NS}/environment/float_cam/float_cam",
              update_period=0.1,
              height=640,
              width=640,
              data_types=["rgb", "depth"],
              spawn=None,
          )
      
          # lights
          dome_light = AssetBaseCfg(
              prim_path="/World/light",
              spawn=sim_utils.DomeLightCfg(intensity=3000.0, color=(0.75, 0.75, 0.75)),
          )
      
          # object_a
          object_a = RigidObjectCfg(
              prim_path="/World/envs/env_.*/object_a",
              spawn=sim_utils.SphereCfg(
                  radius=1,
                  visual_material=sim_utils.PreviewSurfaceCfg(
                      diffuse_color=(1.0, 0.0, 0.0), metallic=0.2
                  ),
                  rigid_props=sim_utils.RigidBodyPropertiesCfg(
                      solver_position_iteration_count=4,
                      solver_velocity_iteration_count=0,
                      disable_gravity=True,  # Disable gravity
                  ),
                  mass_props=sim_utils.MassPropertiesCfg(mass=1.0),
                  collision_props=sim_utils.CollisionPropertiesCfg(
                      collision_enabled=False  # Disable collisions
                  ),
              ),
              init_state=RigidObjectCfg.InitialStateCfg(pos=(0.0, -8.0, 5.0)),
          )
      
          def __post_init__(self):
              self.replicate_physics = False
      
              def define_markers() -> VisualizationMarkers:
                  """Define markers with various different shapes."""
                  marker_cfg = VisualizationMarkersCfg(
                      prim_path="/World/Visuals/testMarkers",  # Changed to env-specific path
                      markers={
                          "frame": sim_utils.UsdFileCfg(
                              usd_path=f"{ISAAC_NUCLEUS_DIR}/Props/UIElements/frame_prim.usd",
                              scale=(0.5, 0.5, 0.5),
                          ),
                      },
                  )
                  return VisualizationMarkers(marker_cfg)
      
              my_visualizer = define_markers()
      
              def _add_grid_markers(self):
                  # Define a 3D grid from -12, -12, 0 to 12, 12, 10
                  num_markers_per_axis = 5  # Number of markers along each axis
                  grid_spacing_x = (12.0 - (-12.0)) / (num_markers_per_axis - 1)  # 24 / (n-1)
                  grid_spacing_y = (12.0 - (-12.0)) / (num_markers_per_axis - 1)  # 24 / (n-1)
                  grid_spacing_z = (10.0 - 0.0) / (num_markers_per_axis - 1)  # 10 / (n-1)
      
                  # Create ranges for x, y, z centered around the environment's origin
                  x_range = torch.arange(-12.0, 12.0 + grid_spacing_x / 2, grid_spacing_x)
                  y_range = torch.arange(-12.0, 12.0 + grid_spacing_y / 2, grid_spacing_y)
                  z_range = torch.arange(0.0, 10.0 + grid_spacing_z / 2, grid_spacing_z)
      
                  # Create the 3D grid
                  x_grid, y_grid, z_grid = torch.meshgrid(
                      x_range, y_range, z_range, indexing="ij"
                  )
                  x_grid = x_grid.reshape(-1)
                  y_grid = y_grid.reshape(-1)
                  z_grid = z_grid.reshape(-1)
      
                  # Marker locations
                  marker_locations = torch.stack([x_grid, y_grid, z_grid], dim=1)
                  marker_indices = torch.arange(my_visualizer.num_prototypes).repeat(
                      num_markers_per_axis**3  # Total markers = n^3 for 3D grid
                  )
                  yaw = torch.zeros_like(marker_locations[:, 0])
      
                  # Rotate the markers around the z-axis for visualization
                  marker_orientations = quat_from_angle_axis(
                      yaw, torch.tensor([0.0, 0.0, 1.0])
                  )
                  # Visualize
                  my_visualizer.visualize(
                      marker_locations, marker_orientations, marker_indices=marker_indices
                  )
      
              _add_grid_markers(self)

also on another topic, I was curious how I can add a marker on top of my robot camera in each env where the marker points in the direction the camera is pointing in each step.

**Comment by RandomOakForest** [comment]
Thank you for posting this. You may want to review issues #2717 and #2233 for workarounds. If you still need further help, please share your error log.

**Comment by shiven2001** [comment]
According to issues #2717 it seems that markers can ONLY spawn for global prim and not env prim. I just wanted to use them for debugging so I guess it is fine. I will close this thread.
